File: scripts/run_analysis.py
#!/usr/bin/env python3
"""
Main analysis script for medevac interrater reliability study.
"""

import sys
from pathlib import Path
import pandas as pd
import numpy as np

# Add src to path
PROJECT_ROOT = Path(__file__).parent.parent
sys.path.insert(0, str(PROJECT_ROOT / "src"))

from medevac_interrater.data_loader import load_clean_data
from medevac_interrater.analysis import (
    calculate_percentage_agreement,
    calculate_question_level_metrics,
    calculate_agreement_by_class,
    calculate_confidence_analysis,
)


def main():
    """Run the complete interrater reliability analysis."""
    
    print("=" * 80)
    print("MEDEVAC INTERRATER RELIABILITY ANALYSIS")
    print("=" * 80)
    print()
    
    # Setup paths
    data_dir = PROJECT_ROOT / "data"
    output_dir = PROJECT_ROOT / "output"
    output_dir.mkdir(exist_ok=True)
    
    # Load and clean data
    print("📥 Loading and cleaning data...")
    raw_df, long_df = load_clean_data(data_dir)
    
    print(f"   ✓ Loaded {len(raw_df)} physicians")
    print(f"   ✓ Reshaped to {len(long_df)} physician-vignette pairs")
    print(f"   ✓ {len(long_df['physician_id'].unique())} unique physicians")
    print(f"   ✓ {len(long_df['question'].unique())} unique questions")
    print()
    
    # Save cleaned data
    long_df.to_csv(output_dir / "cleaned_data_long.csv", index=False)
    print(f"💾 Saved cleaned data to: {output_dir / 'cleaned_data_long.csv'}")
    print()
    
    # Overall percentage agreement
    print("=" * 80)
    print("OVERALL AGREEMENT")
    print("=" * 80)
    overall_pa = calculate_percentage_agreement(long_df, question=None)
    print(f"Overall Percentage Agreement: {overall_pa:.3f} ({overall_pa*100:.1f}%)")
    print()
    
    # Question-level metrics
    print("=" * 80)
    print("QUESTION-LEVEL METRICS")
    print("=" * 80)
    question_metrics = calculate_question_level_metrics(long_df)
    question_metrics.to_csv(output_dir / "question_level_metrics.csv", index=False)
    print(question_metrics.to_string(index=False))
    print()
    print(f"💾 Saved to: {output_dir / 'question_level_metrics.csv'}")
    print()
    
    # Agreement by vignette class
    print("=" * 80)
    print("AGREEMENT BY VIGNETTE CLASS")
    print("=" * 80)
    class_metrics = calculate_agreement_by_class(long_df)
    class_metrics.to_csv(output_dir / "class_level_metrics.csv", index=False)
    print(class_metrics.to_string(index=False))
    print()
    print(f"💾 Saved to: {output_dir / 'class_level_metrics.csv'}")
    print()
    
    # Confidence analysis
    print("=" * 80)
    print("CONFIDENCE ANALYSIS")
    print("=" * 80)
    conf_by_decision, conf_by_class = calculate_confidence_analysis(long_df)
    
    print("\nConfidence by Decision:")
    print(conf_by_decision.to_string(index=False))
    conf_by_decision.to_csv(output_dir / "confidence_by_decision.csv", index=False)
    
    print("\nConfidence by Vignette Class:")
    print(conf_by_class.to_string(index=False))
    conf_by_class.to_csv(output_dir / "confidence_by_class.csv", index=False)
    print()
    print(f"💾 Saved to: {output_dir / 'confidence_by_decision.csv'}")
    print(f"💾 Saved to: {output_dir / 'confidence_by_class.csv'}")
    print()
    
    # Summary statistics
    print("=" * 80)
    print("SUMMARY STATISTICS")
    print("=" * 80)
    print(f"Mean Percentage Agreement: {question_metrics['percentage_agreement'].mean():.3f}")
    print(f"Mean Fleiss' Kappa: {question_metrics['fleiss_kappa'].mean():.3f}")
    print(f"Mean Confidence: {long_df['confidence'].mean():.2f}")
    print(f"SD Confidence: {long_df['confidence'].std():.2f}")
    print()
    
    # Decision distribution
    print("Decision Distribution:")
    decision_dist = long_df["decision"].value_counts()
    for decision, count in decision_dist.items():
        pct = (count / len(long_df)) * 100
        print(f"  {decision}: {count} ({pct:.1f}%)")
    print()
    
    print("=" * 80)
    print("✅ Analysis complete!")
    print("=" * 80)
    print(f"\nAll results saved to: {output_dir}")
    print()


if __name__ == "__main__":
    main()

File: scripts/r/run_analysis.R
#!/usr/bin/env Rscript
# Main R analysis script for medevac interrater reliability study
#
# This script loads processed data from Python and performs all statistical analyses.

# Setup - find project root
# Try to find project root by looking for common markers
find_project_root <- function() {
  current <- normalizePath(getwd())
  max_depth <- 10
  depth <- 0
  
  while (depth < max_depth) {
    if (file.exists(file.path(current, "data", "processed")) ||
        file.exists(file.path(current, "R", "R", "analysis.R"))) {
      return(current)
    }
    parent <- dirname(current)
    if (parent == current) break
    current <- parent
    depth <- depth + 1
  }
  
  # Fallback: assume we're in scripts/r/
  return(normalizePath(file.path(getwd(), "..", "..")))
}

project_root <- find_project_root()

# Source analysis functions
source(file.path(project_root, "R", "R", "analysis.R"))

# Paths
data_dir <- file.path(project_root, "data", "processed")
output_dir <- file.path(project_root, "output")
dir.create(output_dir, showWarnings = FALSE)

# Load processed data
cat(strrep("=", 80), "\n")
cat("MEDEVAC INTERRATER RELIABILITY ANALYSIS - R\n")
cat(strrep("=", 80), "\n\n")

data_file <- file.path(data_dir, "survey_data_processed.csv")
if (!file.exists(data_file)) {
  stop("Processed data file not found. Please run Python data processing first.")
}

cat("📥 Loading processed data...\n")
data <- read.csv(data_file, stringsAsFactors = FALSE)
cat(sprintf("   ✓ Loaded %d rows\n", nrow(data)))
cat(sprintf("   ✓ %d unique physicians\n", length(unique(data$physician_id))))
cat(sprintf("   ✓ %d unique questions\n", length(unique(data$question))))
cat("\n")

# Overall percentage agreement
cat(strrep("=", 80), "\n")
cat("OVERALL AGREEMENT\n")
cat(strrep("=", 80), "\n")
overall_pa <- calculate_percentage_agreement(data, question_id = NULL)
cat(sprintf("Overall Percentage Agreement: %.3f (%.1f%%)\n", overall_pa, overall_pa * 100))
cat("\n")

# Question-level metrics
cat(strrep("=", 80), "\n")
cat("QUESTION-LEVEL METRICS\n")
cat(strrep("=", 80), "\n")
question_metrics <- calculate_question_level_metrics(data)
print(question_metrics)
write.csv(question_metrics, 
          file.path(output_dir, "question_level_metrics.csv"), 
          row.names = FALSE)
cat("\n💾 Saved to: question_level_metrics.csv\n\n")

# Agreement by vignette class
cat(strrep("=", 80), "\n")
cat("AGREEMENT BY VIGNETTE CLASS\n")
cat(strrep("=", 80), "\n")
class_metrics <- calculate_agreement_by_class(data)
print(class_metrics)
write.csv(class_metrics, 
          file.path(output_dir, "class_level_metrics.csv"), 
          row.names = FALSE)
cat("\n💾 Saved to: class_level_metrics.csv\n\n")

# Confidence analysis
cat(strrep("=", 80), "\n")
cat("CONFIDENCE ANALYSIS\n")
cat(strrep("=", 80), "\n")
conf_analysis <- analyze_confidence(data)
cat("\nConfidence by Decision:\n")
print(conf_analysis$by_decision)
write.csv(conf_analysis$by_decision, 
          file.path(output_dir, "confidence_by_decision.csv"), 
          row.names = FALSE)

cat("\nConfidence by Vignette Class:\n")
print(conf_analysis$by_class)
write.csv(conf_analysis$by_class, 
          file.path(output_dir, "confidence_by_class.csv"), 
          row.names = FALSE)
cat("\n💾 Saved confidence analyses\n\n")

# Summary statistics
cat(strrep("=", 80), "\n")
cat("SUMMARY STATISTICS\n")
cat(strrep("=", 80), "\n")
cat(sprintf("Mean Percentage Agreement: %.3f\n", mean(question_metrics$percentage_agreement, na.rm = TRUE)))
cat(sprintf("Mean Fleiss' Kappa: %.3f\n", mean(question_metrics$fleiss_kappa, na.rm = TRUE)))
cat(sprintf("Mean Confidence: %.2f\n", mean(data$confidence, na.rm = TRUE)))
cat(sprintf("SD Confidence: %.2f\n", sd(data$confidence, na.rm = TRUE)))
cat("\n")

# Decision distribution
cat("Decision Distribution:\n")
decision_dist <- table(data$decision)
for (decision in names(decision_dist)) {
  count <- decision_dist[[decision]]
  pct <- (count / nrow(data)) * 100
  cat(sprintf("  %s: %d (%.1f%%)\n", decision, count, pct))
}
cat("\n")

cat(strrep("=", 80), "\n")
cat("✅ Analysis complete!\n")
cat(strrep("=", 80), "\n")
cat(sprintf("\nAll results saved to: %s\n", output_dir))
cat("\n")

